### DISCUSSION TITLE
Avoid Init for changing IV - Realtime use case

### DISCUSSION THREAD

**Discussion by wtedeschi** [openssl/openssl#22161] [Q&A] [answered]
Dear Crypto Experts,

We are looking for the possibility to speedup the encryption process.
As a context the application we are talking about is a realtime communication.
The sequence of calls we are using is:
EVP_CIPHER_CTX_reset
EVP_EncryptInit_ex2
EVP_EncryptUpdate
EVP_EncryptFinal_ex
EVP_CIPHER_CTX_ctrl

On the real time communication several (hundreds) of frames are send per cycle. Every frame will have its own IV. This way we are looking for a API to introduce the IV after the init function, avoiding the overhead of the init for every frame to be transmitted, is that possible?
Would it make sense to enter a feature request for that?

Thanks a lot for your feedback,
Fred

**Comment by t8m** [comment, marked as answer]
You can call EVP_EncryptInit_ex2() with just the iv set and the cipher + key parameters being NULL. This has as low overhead as possible as it will reuse the key schedule from the previous EncryptInit call.

**Reply by wtedeschi** [reply]
Thanks for the quick answer! 
I have tried with the sample I had, benchmark was really good, like 1/3 of the time. But in terms of functionality I failed to decrypt it. Probably I am doing something wrong. 
Is there any code/test for that where we could take a look on the code? 
Thanks again!

**Reply by t8m** [reply]
There is some example of reinitialization in test_cipher_reinit() in test/evp_libctx_test.c
